icu4c/source/test/intltest/locnmtst.cpp

```diff
@@ -422,4 +424,40 @@ void LocaleDisplayNamesTest::TestRootEtc() {
   delete ldn;
 }
 
+void LocaleDisplayNamesTest::TestUndDisplayNames() {
+    IcuTestErrorCode status(*this, "TestUndDisplayNames()");
+
+    static const Locale undWithoutRegion = Locale::forLanguageTag("und", status);
+    static const Locale undWithRegion = Locale::forLanguageTag("und-CH", status);
+
+    LocalPointer<LocaleDisplayNames> namesWithoutRegion(LocaleDisplayNames::createInstance(undWithoutRegion));
+    LocalPointer<LocaleDisplayNames> namesWithRegion(LocaleDisplayNames::createInstance(undWithRegion));
+
+    UnicodeString result;
+
+    result.remove();
+    namesWithoutRegion->languageDisplayName("", result);
+    test_assert_equal("Unknown language", result);
+
+    result.remove();
+    namesWithoutRegion->languageDisplayName("und", result);
+    test_assert_equal("Unknown language", result);
+
+    result.remove();
+    namesWithoutRegion->regionDisplayName("CH", result);
+    test_assert_equal("Switzerland", result);
+
+    result.remove();
+    namesWithRegion->languageDisplayName("", result);
+    test_assert_equal("Unknown language", result);
```

Comment:
The _CH bundle lookup should go to the default locale (inside ures_open()), so the results here should depend on what the default locale is -- maybe English, maybe not. For testing, we should probably fetch the strings for Locale::getDefault() and use those as expected results.